src/particles/diagnostics/CovarianceMatrixMath.H

```diff
@@ -0,0 +1,209 @@
+     *
+     * @param[in] a coefficient of cubic term
+     * @param[in] b coefficient of quadratic term
+     * @param[in] c coefficient of linear term
+     * @param[in] d coefficient of constant term
+     * @returns tuple of three real roots
+     */
+    std::tuple<
+            amrex::ParticleReal,
+            amrex::ParticleReal,
+            amrex::ParticleReal>
+    CubicRootsAlg (
+        amrex::ParticleReal a,
+        amrex::ParticleReal b,
+        amrex::ParticleReal c,
+        amrex::ParticleReal d
+    )
+    {
+        using namespace amrex::literals;
+        using Complex = amrex::GpuComplex<amrex::ParticleReal>;
+
+        std::tuple <amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal> roots;
+        amrex::ParticleReal x1 = 0.0_prt;
+        amrex::ParticleReal x2 = 0.0_prt;
+        amrex::ParticleReal x3 = 0.0_prt;
+
+        amrex::ParticleReal Q = (3.0_prt*a*c - std::pow(b,2))/(9.0_prt * std::pow(a,2));
+        amrex::ParticleReal R = (9.0_prt*a*b*c - 27_prt*std::pow(a,2)*d - 2.0_prt*std::pow(b,3))/(54.0_prt*std::pow(a,3));
+        amrex::ParticleReal discriminant = std::pow(Q,3) + std::pow(R,2);
+
+        // Define complex variable C
+        Complex Qc(Q,0.0_prt);
+        Complex Rc(R,0.0_prt);
+        Complex Dc(discriminant,0.0_prt);
+        Complex C = amrex::pow(-Rc + amrex::sqrt(Dc),1.0/3.0);
+
+        // Define a cubic root of unity xi
+        amrex::ParticleReal xire = -1.0/2.0;
+        amrex::ParticleReal xiim = std::sqrt(3.0)/2.0;
+        Complex xi(xire,xiim);
+
+        //Three roots in algebraic form.
+
+        if (C.m_real == 0.0_prt && C.m_imag == 0.0_prt){  // Special case of a triple root
+
+           x1 = - b/(3.0_prt*a);
+           x2 = - b/(3.0_prt*a);
+           x3 = - b/(3.0_prt*a);
+
+        } else {
+
+           Complex z1 = Qc/C - C;
+           Complex z2 = Qc/(xi*C) - xi*C;
+           Complex z3 = Qc/(amrex::pow(xi,2)*C) - amrex::pow(xi,2)*C;
+           x1 = z2.m_real - b/(3.0*a);
+           x2 = z1.m_real - b/(3.0*a);
+           x3 = z3.m_real - b/(3.0*a);
+        }
+
+
+        roots = std::make_tuple(x1,x2,x3);
+        return roots;
+    }
+
+
+    /** Function to take the trace of a square 6x6 matrix.
+     *
+     * @param[in] A a square matrix
+     * @returns the trace of A
+     */
+    amrex::ParticleReal
+    TraceMat (
+        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> A
+    )
+    {
+        int const dim = 6;
+        amrex::ParticleReal trA = 0.0;
+
+        for (int i = 1; i < dim+1; i++) {
+            trA += A(i,i);
+        }
+        return trA;
+    }
+
+
+    /** Function to multiply two square matrices of dimension 6.
+     *
+     * @param[in] A a square matrix
+     * @param[in] B square matrix
+     * @returns the matrix C = AB
+     */
+    amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6>
+    MultiplyMat (
+        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> A,
+        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> B
```

Comment:
Avoid copying A and B:
```suggestion
        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> const & A,
        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> const & B
```